File: packages/js-dpp/lib/errors/DocumentTimestampsMismatchError.js
const ConsensusError = require('./ConsensusError');

class DocumentTimestampsMismatchError extends ConsensusError {
  /**
   * @param {DocumentCreateTransition} documentTransition
   */
  constructor(documentTransition) {
    super('Document createdAt and updatedAt timestamps are not equal');

    this.documentTransition = documentTransition;
  }

  /**
   * Get document create transition
   *
   * @return {DocumentCreateTransition}
   */
  getDocumentTransition() {
    return this.documentTransition;
  }
}

module.exports = DocumentTimestampsMismatchError;

File: packages/rs-dpp/src/errors/consensus/state/document/document_timestamps_mismatch_error.rs
use crate::errors::consensus::state::state_error::StateError;
use crate::errors::consensus::ConsensusError;
use crate::errors::ProtocolError;
use bincode::{Decode, Encode};
use platform_serialization_derive::{PlatformDeserialize, PlatformSerialize};
use platform_value::Identifier;
use thiserror::Error;

#[derive(
    Error, Debug, Clone, PartialEq, Eq, Encode, Decode, PlatformSerialize, PlatformDeserialize,
)]
#[error("Document {document_id} createdAt and updatedAt timestamps are not equal")]
#[platform_serialize(unversioned)]
#[cfg_attr(feature = "apple", ferment_macro::export)]
pub struct DocumentTimestampsMismatchError {
    /*

    DO NOT CHANGE ORDER OF FIELDS WITHOUT INTRODUCING OF NEW VERSION

    */
    pub document_id: Identifier,
}

impl DocumentTimestampsMismatchError {
    pub fn new(document_id: Identifier) -> Self {
        Self { document_id }
    }

    pub fn document_id(&self) -> &Identifier {
        &self.document_id
    }
}

impl From<DocumentTimestampsMismatchError> for ConsensusError {
    fn from(err: DocumentTimestampsMismatchError) -> Self {
        Self::StateError(StateError::DocumentTimestampsMismatchError(err))
    }
}
